provider/mem/mem.go

```diff
@@ -77,7 +77,7 @@ func NewAlerts(ctx context.Context, m types.Marker, intervalGC time.Duration, al
 		logger:    log.With(l, "component", "provider"),
 		callback:  alertCallback,
 	}
-	a.alerts.SetGCCallback(func(alerts []*types.Alert) {
+	a.alerts.SetResolvedCallback(func(alerts []*types.Alert) {
```

Comment:
A part from `m.Delete(alert.Fingerprint())`, is it correct to execute the rest of the logic for the case this callback is called because an explicit call to `Alerts.Delete()`? Could this lead to any side effect? Have you checked it?

Reply:
I verified that `Delete` is only called from [within dispatch](https://github.com/prometheus/alertmanager/blob/master/dispatch/dispatch.go#L529-L533).

But I think you might be right, given we now have a chance of calling this callback more than once concurrently I should double-check that the operation within it is safe. I'll do that tomorrow and let you know.

Reply:
Thanks for raising this, I've now taken a look. The way I understand it, an explicit call to `Alerts.Delete()` would only occur if (due to a timing issue) the alert would resolve **just before** we're about to send it.

https://github.com/prometheus/alertmanager/blob/fd0929ba9fc58737a9c91f24771862692fa72d17/dispatch/dispatch.go#L529-L533

While (to the best of my knowledge) the deletion of no longer used listeners at a more regular interval (each receiver/inhibitions subscribes to the alerts store) doesn't produce any side effects, I don't think a single alert resolving should trigger the garbage collection of this.

But, there's more... After testing more thoroughly I realised that this doesn't really solve what we wanted in the first place (accurate reporting of the `alertmanager_alerts`) because if an alert is resolved naturally it would be GCed until the next run anyway.

An example of this behaviour (with this change included) is that despite seeing 67 alerts in the UI and the metric reporting 72 alerts, it was not until the GC cycle ran (every 30 minutes by default) that I could see the accurate measurement:

Metrics
![Screenshot 2021-06-23 at 18 18 15](https://user-images.githubusercontent.com/231583/123140767-96a78d00-d44f-11eb-999d-fd29886ed027.png)

API (all filters are enabled by default)
![Screenshot 2021-06-23 at 18 20 07](https://user-images.githubusercontent.com/231583/123140857-b048d480-d44f-11eb-8964-59d56c2f1543.png)

Given there's no way to see resolved alerts through the UI, My proposal now is that we actually create a new label of "resolved" for this metric which would allow us to accurately track the value over time in accordance with what we see in the UI. 


With that we can do 

```promQL
sum(alertmanager_alerts{state=~"active|supressed"}) - sum(alertmanager_alerts{state="resolved"})
```
to match what we see in the UI.